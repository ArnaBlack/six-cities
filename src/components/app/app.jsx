import React, {PureComponent} from 'react';
import Main from '../main/main.jsx';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import ActionCreators from '../../store/actions/action-creators';

class App extends PureComponent {
  render() {
    const {
      currentCity,
      offers,
    } = this.props;

    return <Main
      currentCity={currentCity}
      offers={offers}
    />;
  }

  componentDidMount() {
    const {getOffers} = this.props;
    getOffers();
  }
}

App.propTypes = {
  currentCity: PropTypes.string.isRequired,
  offers: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired,
    mark: PropTypes.string,
    imageSrc: PropTypes.string.isRequired,
    price: PropTypes.number.isRequired,
    inBookmarks: PropTypes.bool.isRequired,
    rating: PropTypes.number.isRequired,
    title: PropTypes.string.isRequired,
    type: PropTypes.string.isRequired,
    coordinates: PropTypes.arrayOf(PropTypes.number),
  })),
  getOffers: PropTypes.func.isRequired,
};

App.defaultProps = {
  offers: [],
};

const mapStateToProps = (state, props) => ({
  ...props,
  currentCity: state.city,
  offers: state.offers.filter((it) => it.city === state.city),
});

const mapDispatchToProps = (dispatch) => ({
  getOffers: () => dispatch(ActionCreators.getOffers())
});

export {App};
export default connect(mapStateToProps, mapDispatchToProps)(App);
