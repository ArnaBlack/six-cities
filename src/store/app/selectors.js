import NameSpace from '../name-space';

export const getCurrentCity = (state) => state[NameSpace.APP].city;
