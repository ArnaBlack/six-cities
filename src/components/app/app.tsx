import * as React from 'react';
import {connect} from 'react-redux';
import {
  Switch,
  Route,
  Redirect,
} from 'react-router-dom';

import withActiveItem from '../../hocs/with-active-item/with-active-item';
import withTransformProps from '../../hocs/with-transform-props/with-transform-props';
import withPrivateRoute from '../../hocs/with-private-route/with-private-route';

import Loader from '../loader/loader';
import MainPage from '../main-page/main-page';
import SignInPage from '../sign-in-page/sign-in-page';
import FavoritesPage from '../favorites-page/favorites-page';
import PlacePage from '../place-page/place-page';

import UserOperation from '../../store/user/operation/operation';
import OffersOperation from '../../store/offers/operation/operation';
import {getLoadingState} from '../../store/offers/selectors';
import {getAuthorizationStatus} from '../../store/user/selectors';

import {City} from '../../types';

const transformActiveToSelected = (props) => ({
  selectedOffer: props.activeItem,
  onSelectOffer: props.onSelectItem,
});
const MainPageWrapped = withActiveItem(withTransformProps(transformActiveToSelected)(MainPage));
const FavoritesPageWrapped = withPrivateRoute(FavoritesPage);

interface Props {
  isLoading: boolean,
  currentCity: City,
  isAuthorizationRequired: boolean,
  checkAuth: () => void,
  loadOffers: () => void,
}

class App extends React.PureComponent<Props, null> {
  render() {
    const {
      isLoading,
      isAuthorizationRequired,
    } = this.props;

    return isLoading ? <Loader /> : <Switch>
      <Route path="/" exact component={MainPageWrapped} />
      <Route path="/login" render={() => isAuthorizationRequired ? <SignInPage /> : <Redirect to="/" />}/>
      <Route path="/favorites" component={FavoritesPageWrapped} />
      <Route path="/offer/:id" component={PlacePage} />
      <Redirect from="*" to="/" />
    </Switch>;
  }

  componentDidMount() {
    const {
      checkAuth,
      loadOffers,
    } = this.props;

    checkAuth();
    loadOffers();
  }
}

const mapStateToProps = (state, props) => ({
  ...props,
  isLoading: getLoadingState(state),
  isAuthorizationRequired: getAuthorizationStatus(state),
});

const mapDispatchToProps = (dispatch) => ({
  checkAuth: () => dispatch(UserOperation.checkAuth()),
  loadOffers: () => dispatch(OffersOperation.loadOffers()),
});

export {App};
export default connect(mapStateToProps, mapDispatchToProps)(App);
