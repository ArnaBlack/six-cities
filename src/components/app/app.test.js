import React from 'react';
import Enzyme, {shallow} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import {shallowToJson} from 'enzyme-to-json';
import {App} from './app.jsx';

Enzyme.configure({adapter: new Adapter()});

const mock = {
  currentCity: `Amsterdam`,
  cityCoordinates: [52.38333, 4.9],
  offers: [
    {
      id: 1,
      mark: `Premium`,
      imageSrc: `img/apartment-01.jpg`,
      price: 120,
      inBookmarks: true,
      rating: 4,
      title: `Beautiful luxurious apartment at great location`,
      type: `Apartment`,
      coordinates: [52.3909553943508, 4.85309666406198],
    },
  ],
};

it(`App correctly renders`, () => {
  const {
    currentCity,
    cityCoordinates,
    offers
  } = mock;
  const tree = shallow(<App
    currentCity={currentCity}
    cityCoordinates={cityCoordinates}
    offers={offers}
    getOffers={jest.fn()}
  />);

  expect(shallowToJson(tree)).toMatchSnapshot();
});
