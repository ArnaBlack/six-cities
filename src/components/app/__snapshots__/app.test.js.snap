// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App correctly renders 1`] = `
<Main
  currentCity="Amsterdam"
  offers={
    Array [
      Object {
        "coordinates": Array [
          52.3909553943508,
          4.85309666406198,
        ],
        "id": 1,
        "imageSrc": "img/apartment-01.jpg",
        "inBookmarks": true,
        "mark": "Premium",
        "price": 120,
        "rating": 4,
        "title": "Beautiful luxurious apartment at great location",
        "type": "Apartment",
      },
    ]
  }
/>
`;
